### PR TITLE
feat: add support for `http_interface` and `http_bind_address`

### PR DESCRIPTION
**Pull request by tenthirtyam**
### Summary

Adds support for `http_interface` and `http_bind_interface` from the SDK.

> [!NOTE] 
> - The options `http_bind_address` and `http_interface` are mutually exclusive, per the SDK.
> - Both `http_bind_address` and `http_interface` have higher priority than `http_ip`.
> - The `http_bind_address` is matched against the IP addresses of the host's network interfaces. If no match is found, the plugin will terminate.
> - Similarly, `http_interface` is compared with the host's network interfaces. If there's no corresponding network interface, the plugin will also terminate.
> - If neither `http_bind_address`, `http_interface`, and `http_ip` are provided, the plugin will automatically find and use the IP address of the first non-loopback interface for `http_ip`.
> - If `http_ip` is provided and the IP address does not exist on any interface, the build will exit.

### Testing

**Basic**

```sh
packer-plugin-vsphere on  feat/http-interface [!] via 🐹 v1.22.6 
➜ go fmt ./...

packer-plugin-vsphere on  feat/http-interface [!] via 🐹 v1.22.6 
➜ make generate
2024/08/09 10:23:36 Copying "docs" to ".docs/"
2024/08/09 10:23:36 Replacing @include '...' calls in .docs/
Compiling MDX docs in '.docs' to Markdown in '.web-docs'...

packer-plugin-vsphere on  feat/http-interface [!] via 🐹 v1.22.6 took 16.6s 
➜ make build

packer-plugin-vsphere on  feat/http-interface [!] via 🐹 v1.22.6 took 6.5s 
➜ make test

?       github.com/hashicorp/packer-plugin-vsphere      [no test files]
?       github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/common/testing       [no test files]
ok      github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/clone        1.664s
?       github.com/hashicorp/packer-plugin-vsphere/examples/driver      [no test files]
?       github.com/hashicorp/packer-plugin-vsphere/version      [no test files]
ok      github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/common       3.108s
ok      github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/driver       6.676s
ok      github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/iso  2.554s
ok      github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/supervisor   5.676s
ok      github.com/hashicorp/packer-plugin-vsphere/post-processor/vsphere       1.746s
ok      github.com/hashicorp/packer-plugin-vsphere/post-processor/vsphere-template      3.037s

packer-plugin-vsphere took 1m 3.7s …
➜ 

packer-plugin-vsphere on  feat/http-interface [!] via 🐹 v1.22.6 
➜ make dev                       
packer plugins install --path packer-plugin-vsphere "github.com/hashicorp/vsphere"
Successfully installed plugin github.com/hashicorp/vsphere from /Users/ryan/Library/Mobile Documents/com~apple~CloudDocs/Code/Personal/packer-plugin-vsphere/packer-plugin-vsphere to /Users/ryan/.packer.d/plugins/github.com/hashicorp/vsphere/packer-plugin-vsphere_v1.3.1-dev_x5.0_darwin_arm64

```

✅  **PASS** : `http_ip` not provided. (Default)

```sh
==> vsphere-iso.appliance: Starting HTTP server on port 8091
==> vsphere-iso.appliance: Setting boot order...
==> vsphere-iso.appliance: Powering on virtual machine...
==> vsphere-iso.appliance: Waiting 2s for boot...
==> vsphere-iso.appliance: Serving HTTP requests at http://192.168.86.58:8091/.
==> vsphere-iso.appliance: Typing boot command...
```

```log
2024/08/09 10:30:06 ui: [1;32m==> vsphere-iso.appliance: Starting HTTP server on port 8950[0m
2024/08/09 10:30:06 ui: [1;32m==> vsphere-iso.appliance: Setting boot order...[0m
2024/08/09 10:30:06 ui: [1;32m==> vsphere-iso.appliance: Powering on virtual machine...[0m
2024/08/09 10:30:07 ui: [1;32m==> vsphere-iso.appliance: Waiting 2s for boot...[0m
2024/08/09 10:30:09 packer-plugin-vsphere_v1.3.1-dev_x5.0_darwin_arm64 plugin: 2024/08/09 10:30:09 Using IP address 192.168.86.58 from http_ip.
2024/08/09 10:30:09 ui: [1;32m==> vsphere-iso.appliance: Serving HTTP requests at http://192.168.86.58:8950/.[0m
2024/08/09 10:30:09 ui: [1;32m==> vsphere-iso.appliance: Typing boot command...[0m
```

✅  **PASS** : `http_ip` provided with a **valid** IP address.

```sh
==> vsphere-iso.appliance: Starting HTTP server on port 8457
==> vsphere-iso.appliance: Setting boot order...
==> vsphere-iso.appliance: Powering on virtual machine...
==> vsphere-iso.appliance: Waiting 2s for boot...
==> vsphere-iso.appliance: Serving HTTP requests at http://192.168.86.58:8457/.
==> vsphere-iso.appliance: Typing boot command...
```

```log
2024/08/09 10:31:21 ui: [1;32m==> vsphere-iso.appliance: Starting HTTP server on port 8457[0m
2024/08/09 10:31:21 ui: [1;32m==> vsphere-iso.appliance: Setting boot order...[0m
2024/08/09 10:31:21 ui: [1;32m==> vsphere-iso.appliance: Powering on virtual machine...[0m
2024/08/09 10:31:21 ui: [1;32m==> vsphere-iso.appliance: Waiting 2s for boot...[0m
2024/08/09 10:31:23 packer-plugin-vsphere_v1.3.1-dev_x5.0_darwin_arm64 plugin: 2024/08/09 10:31:23 Using IP address 192.168.86.58 from http_ip.
2024/08/09 10:31:23 ui: [1;32m==> vsphere-iso.appliance: Serving HTTP requests at http://192.168.86.58:8457/.[0m
2024/08/09 10:31:23 ui: [1;32m==> vsphere-iso.appliance: Typing boot command...[0m
```

✅  **PASS** : `http_ip` provided with an **invalid** IP address.

```sh
--> vsphere-iso.appliance: error using IP address 192.168.86.57: 192.168.86.57 is not assigned to an interface
```

```log
2024/08/09 10:32:14 ui error: [1;31m==> vsphere-iso.appliance: error using IP address 192.168.86.57: 192.168.86.57 is not assigned to an interface[0m
```

✅  **PASS** : `http_bind_address` provided with matching interface.

```sh
==> vsphere-iso.appliance: Starting HTTP server on port 8996
==> vsphere-iso.appliance: Setting boot order...
==> vsphere-iso.appliance: Powering on virtual machine...
==> vsphere-iso.appliance: Waiting 2s for boot...
==> vsphere-iso.appliance: Serving HTTP requests at http://192.168.86.58:8996/.
==> vsphere-iso.appliance: Typing boot command...
```

```log
2024/08/09 10:36:50 ui: [1;32m==> vsphere-iso.appliance: Starting HTTP server on port 8996[0m
2024/08/09 10:36:50 ui: [1;32m==> vsphere-iso.appliance: Setting boot order...[0m
2024/08/09 10:36:51 ui: [1;32m==> vsphere-iso.appliance: Powering on virtual machine...[0m
2024/08/09 10:36:51 ui: [1;32m==> vsphere-iso.appliance: Waiting 2s for boot...[0m
2024/08/09 10:36:53 packer-plugin-vsphere_v1.3.1-dev_x5.0_darwin_arm64 plugin: 2024/08/09 10:36:53 Using IP address 192.168.86.58 from http_bind_address.
2024/08/09 10:36:53 ui: [1;32m==> vsphere-iso.appliance: Serving HTTP requests at http://192.168.86.58:8996/.[0m
2024/08/09 10:36:53 ui: [1;32m==> vsphere-iso.appliance: Typing boot command...[0m
```

✅  **PASS** : `http_bind_address` provided with non-matching interface.

```sh
--> vsphere-iso.appliance: 192.168.86.57 is not assigned to an interface
```

```log
2024/08/09 10:35:33 ui error: [1;31m==> vsphere-iso.appliance: error validating IP address for HTTP server: 192.168.86.57 is not assigned to an interface[0m
```

✅  **PASS** : `http_bind_address` provided with `http_interface`.

```sh
Error: 1 error(s) occurred:

* either http_interface or http_bind_address can be specified
```

```log
2024/08/09 10:37:58 ui error: Error: 1 error(s) occurred:

* either http_interface or http_bind_address can be specified
```

✅  **PASS** : `http_bind_address` provided with `http_ip`. `http_bind_address` takes precedence.

```sh
==> vsphere-iso.appliance: Starting HTTP server on port 8665
==> vsphere-iso.appliance: Setting boot order...
==> vsphere-iso.appliance: Powering on virtual machine...
==> vsphere-iso.appliance: Waiting 2s for boot...
==> vsphere-iso.appliance: Serving HTTP requests at http://192.168.86.58:8665/.
==> vsphere-iso.appliance: Typing boot command...
```

```log
2024/08/09 10:39:17 ui: [1;32m==> vsphere-iso.appliance: Starting HTTP server on port 8665[0m
2024/08/09 10:39:17 ui: [1;32m==> vsphere-iso.appliance: Setting boot order...[0m
2024/08/09 10:39:17 ui: [1;32m==> vsphere-iso.appliance: Powering on virtual machine...[0m
2024/08/09 10:39:18 ui: [1;32m==> vsphere-iso.appliance: Waiting 2s for boot...[0m
2024/08/09 10:39:20 packer-plugin-vsphere_v1.3.1-dev_x5.0_darwin_arm64 plugin: 2024/08/09 10:39:20 Using IP address 192.168.86.58 from http_bind_address.
2024/08/09 10:39:20 ui: [1;32m==> vsphere-iso.appliance: Serving HTTP requests at http://192.168.86.58:8665/.[0m
2024/08/09 10:39:20 ui: [1;32m==> vsphere-iso.appliance: Typing boot command...[0m
```

✅  **PASS** : `http_interface` provided.

🟢 Interface `en0`:

```sh
==> vsphere-iso.appliance: Starting HTTP server on port 8259
==> vsphere-iso.appliance: Setting boot order...
==> vsphere-iso.appliance: Powering on virtual machine...
==> vsphere-iso.appliance: Waiting 2s for boot...
==> vsphere-iso.appliance: Serving HTTP requests at http://192.168.86.58:8259/.
==> vsphere-iso.appliance: Typing boot command...
```

```log
2024/08/09 10:40:41 ui: [1;32m==> vsphere-iso.appliance: Starting HTTP server on port 8259[0m
2024/08/09 10:40:41 packer-plugin-vsphere_v1.3.1-dev_x5.0_darwin_arm64 plugin: 2024/08/09 10:40:41 Found available port: 8259 on IP: 0.0.0.0
2024/08/09 10:40:41 ui: [1;32m==> vsphere-iso.appliance: Setting boot order...[0m
2024/08/09 10:40:41 ui: [1;32m==> vsphere-iso.appliance: Powering on virtual machine...[0m
2024/08/09 10:40:41 ui: [1;32m==> vsphere-iso.appliance: Waiting 2s for boot...[0m
2024/08/09 10:40:43 packer-plugin-vsphere_v1.3.1-dev_x5.0_darwin_arm64 plugin: 2024/08/09 10:40:43 Using IP address 192.168.86.58 from http_interface en0.
2024/08/09 10:40:43 ui: [1;32m==> vsphere-iso.appliance: Serving HTTP requests at http://192.168.86.58:8259/.[0m
2024/08/09 10:40:43 ui: [1;32m==> vsphere-iso.appliance: Typing boot command...[0m
```

🟢 Interface `en1`:

```sh
==> vsphere-iso.appliance: Starting HTTP server on port 8186
==> vsphere-iso.appliance: Setting boot order...
==> vsphere-iso.appliance: Powering on virtual machine...
==> vsphere-iso.appliance: Waiting 2s for boot...
==> vsphere-iso.appliance: Serving HTTP requests at http://192.168.87.57:8186/.
==> vsphere-iso.appliance: Typing boot command...
```

```log
2024/08/09 10:41:24 ui: [1;32m==> vsphere-iso.appliance: Starting HTTP server on port 8186[0m
2024/08/09 10:41:24 ui: [1;32m==> vsphere-iso.appliance: Setting boot order...[0m
2024/08/09 10:41:24 ui: [1;32m==> vsphere-iso.appliance: Powering on virtual machine...[0m
2024/08/09 10:41:24 ui: [1;32m==> vsphere-iso.appliance: Waiting 2s for boot...[0m
2024/08/09 10:41:26 ui: [1;32m==> vsphere-iso.appliance: Serving HTTP requests at http://192.168.213.1:8186/.[0m
2024/08/09 10:41:26 packer-plugin-vsphere_v1.3.1-dev_x5.0_darwin_arm64 plugin: 2024/08/09 10:41:26 Using IP address 192.168.87.57 from http_interface en1.
2024/08/09 10:41:26 ui: [1;32m==> vsphere-iso.appliance: Typing boot command...[0m
```

✅  **PASS** : `http_interface` provided with no matching interface.

```sh
--> vsphere-iso.appliance: error using interface foo: route ip+net: no such network interface
```

```log
2024/08/09 10:43:02 ui error: [1;31mBuild 'vsphere-iso.appliance' errored after 3 seconds 808 milliseconds: error using interface foo: route ip+net: no such network interface[0m
```

✅  **PASS** : `http_interface` provided with `http_ip`. `http_interface` takes precedence.

```sh
==> vsphere-iso.appliance: Starting HTTP server on port 8564
==> vsphere-iso.appliance: Setting boot order...
==> vsphere-iso.appliance: Powering on virtual machine...
==> vsphere-iso.appliance: Waiting 2s for boot...
==> vsphere-iso.appliance: Serving HTTP requests at http://192.168.86.58:8564/.
==> vsphere-iso.appliance: Typing boot command...
```

```log
2024/08/09 10:44:23 ui: [1;32m==> vsphere-iso.appliance: Starting HTTP server on port 8564[0m
2024/08/09 10:44:23 ui: [1;32m==> vsphere-iso.appliance: Setting boot order...[0m
2024/08/09 10:44:23 ui: [1;32m==> vsphere-iso.appliance: Powering on virtual machine...[0m
2024/08/09 10:44:23 ui: [1;32m==> vsphere-iso.appliance: Waiting 2s for boot...[0m
2024/08/09 10:44:25 packer-plugin-vsphere_v1.3.1-dev_x5.0_darwin_arm64 plugin: 2024/08/09 10:44:25 Using IP address 192.168.86.58 from http_interface en0.
2024/08/09 10:44:25 ui: [1;32m==> vsphere-iso.appliance: Serving HTTP requests at http://192.168.86.58:8564/.[0m
2024/08/09 10:44:25 ui: [1;32m==> vsphere-iso.appliance: Typing boot command...[0m
```

### CODE REVIEW

File: builder/vsphere/clone/builder.go
Comment (lbajolet-hashicorp) [inline builder/vsphere/clone/builder.go:99]: It looks like this code is essentially the same between iso/clone; any chance we can factorise it?

**Review by lbajolet-hashicorp — COMMENTED** [review]
Just a couple of comments on the new code, but overall LGTM, thanks for taking the time to delve into this @tenthirtyam

File: builder/vsphere/common/step_http_ip_discover.go
Comment (tenthirtyam) [reply builder/vsphere/common/step_http_ip_discover.go]: Nice catch - much simpler!

Resolved in 8d5fd53.

File: builder/vsphere/common/step_boot_command.go
Comment (tenthirtyam) [reply builder/vsphere/common/step_boot_command.go:99]: Will plan to do a seperate change to update all instances for `ui.Say(fmt.Sprintf` to `ui.Sayf` and `ui.Error(fmt.Sprintf` to `ui.Errorf`.

File: builder/vsphere/clone/builder.go
Comment (tenthirtyam) [reply builder/vsphere/clone/builder.go]: Refactored in 773410c. Retesting pending.

**Comment by tenthirtyam** [comment]
@nywilken This one is ready now, but need some help with the failing est on `go generate`. each of the tests in the **Testing** section still hold true after making factorizing changes Lucas requested.

**Comment by nywilken** [comment]
> @nywilken This one is ready now, but need some help with the failing est on `go generate`. each of the tests in the **Testing** section still hold true after making factorizing changes Lucas requested.

So looking into the failure it looks like the .web-docs were committed to this branch with a previous version of the packer-sdc. Rebasing onto the latest main doesn't fix it because your changes are being applied on top of what is on main. So in order to fix this issue you must run make generate again and commit the updated .web-doc files.

**Comment by tenthirtyam** [comment]
Roget that - will make it so.

File: builder/vsphere/common/step_boot_command.go
Comment (nywilken) [inline builder/vsphere/common/step_boot_command.go:107]: Is it really necessary to read "http_ip_set_by_user" from state and display a custom message if using the default loopback. I was a little confused by this extra check because I would assume if `state.Get("http_ip_set_by_user").(bool)` is true then we know that `http_ip` is being used so this case can be a guard clause above the switch statement. 

Below is a suggested rewrite for the for loop. 
```go
	keys := []string{"http_bind_address", "http_interface", "http_ip"}
	for _, key := range keys {
		value, ok := state.Get(key).(string)
		if ok && value == "" {
			continue
		}

		switch key {
		case "http_bind_address":
			ip = value
			ui.Sayf("Using IP address %s from 'http_bind_address'.", ip)
		case "http_interface":
			ip, err = hostIP(value)
			if err != nil {
				err = fmt.Errorf("error determining IP address from interface %s: %s", value, err)
				state.Put("error", err)
				ui.Error(err.Error())
				return multistep.ActionHalt
			}
			ui.Sayf("Using IP address %s from 'http_interface' %s.", ip, value)
		case "http_ip":
			ip = value
			ui.Sayf("Using IP address %s from 'http_ip'.", ip)
		}
		ui.Sayf("Using IP address %s from %q.", ip, key)
		break
	}
```

File: builder/vsphere/common/step_boot_command.go
Comment (nywilken) [inline builder/vsphere/common/step_boot_command.go:73]: ☝🏽 why is this logic only relevant if port is is greater than 0?

File: builder/vsphere/common/step_boot_command.go
Comment (nywilken) [inline builder/vsphere/common/step_boot_command.go:125]: Since you are outputting the obtained here do you really need to print the IPs address above?

**Review by nywilken — COMMENTED** [review]
I have a few suggestions on the code. But the approach looks good.

File: builder/vsphere/common/step_boot_command.go
Comment (tenthirtyam) [reply builder/vsphere/common/step_boot_command.go:73]: Based on the SDK, this should not be needed because if it's 0 it would not get this far,

However, the SDK does not account for cheking the allowable range? 

- Ports 0 to 1023 are well-known ports, reserved for system or well-known services.
- Port 0 is reserved and cannot be used for normal network communication.
- Registered ports are from 1024 to 49151.
- The remainder of the ports from 49152 to 65535 can be used dynamically by applications.

So I could remove the `if port > 0 {}` and simply allow this to be handled by the SDK or do the following to limit it to between 1024 and 65535

```go
port := state.Get("http_port").(int)
// Check if port is 80 (HTTP reserved) or within the range of 1024 to 65535.
if port == 80 || (port >= 1024 && port <= 65535) {
    ip := state.Get("http_ip").(string)
    s.Ctx.Data = &bootCommandTemplateData{
        IP:   ip,
        Port: port,
        VMName: s.VMName,
    }
    ui.Sayf("HTTP server is working at http://%v:%v/", ip, port)
}
```

File: builder/vsphere/common/step_boot_command.go
Comment (tenthirtyam) [reply builder/vsphere/common/step_boot_command.go:125]: I included the `ui.Say` for the user to see the results, but that could just as easily be logged instead of output to the console.

Comment (tenthirtyam) [reply builder/vsphere/common/step_boot_command.go:128]: Resolved in 53f1531.

File: builder/vsphere/common/step_boot_command.go
Comment (tenthirtyam) [reply builder/vsphere/common/step_boot_command.go:73]: Resolved in 53f1531.

File: builder/vsphere/common/step_boot_command.go
Comment (tenthirtyam) [reply builder/vsphere/common/step_boot_command.go]: Resolved in 53f1531.

File: builder/vsphere/common/http_address.go
Comment (tenthirtyam) [reply builder/vsphere/common/http_address.go]: Resolved in 53f1531.

File: builder/vsphere/clone/builder.go
Comment (tenthirtyam) [reply builder/vsphere/clone/builder.go]: Resolved in 53f1531.

File: builder/vsphere/clone/builder.go
Comment (tenthirtyam) [reply builder/vsphere/clone/builder.go]: Resolved in 53f1531.

File: builder/vsphere/clone/builder.go
Comment (tenthirtyam) [reply builder/vsphere/clone/builder.go]: Resolved in 53f1531.

**Comment by tenthirtyam** [comment]
@nywilken and @lbajolet-hashicorp - this one has been updated but I still need to run another round of testing before any merge. Let me know how the updates look to you both.

File: builder/vsphere/common/step_boot_command.go
Comment (tenthirtyam) [inline builder/vsphere/common/step_boot_command.go:24]: I may move this to `builder/vsphere/common/http_address.go` as `DefaultHttpBindAddress`.

File: builder/vsphere/common/step_boot_command.go
Comment (nywilken) [reply builder/vsphere/common/step_boot_command.go:73]: I'm not sure I understand the issue here. By default, SDK will set the port min to 8000 and port max to 9000. If http_port is 0 an available port between that range will be selected. If you wish to change the port range a users can use `http_port_max` and `http_port_min` in their configuration to change the behavior. 

Please let me know if I am missing something here.

**Review by lbajolet-hashicorp — APPROVED** [review]
Thanks for the rerolls @tenthirtyam, LGTM!

**Comment by tenthirtyam** [comment]
Marked as draft pending e2e testing post refactor.

**Comment by tenthirtyam** [comment]
@lbajolet-hashicorp @nywilken - all comments have been addressed and implemented.

I've tested various scenarios above with the console and log outputs. Note that the log will also show from where the IP address is used.

Ryan

File: builder/vsphere/common/step_boot_command.go
Comment (tenthirtyam) [reply builder/vsphere/common/step_boot_command.go:73]: Ah, so this existed before. It's used as a condition to skip the IP checks for the HTTP server if the port is set.

File: builder/vsphere/common/step_boot_command.go
Comment (lbajolet-hashicorp) [inline builder/vsphere/common/step_boot_command.go:230]: Are we only supporting IPv4 here?

**Review by lbajolet-hashicorp — APPROVED** [review]
Just left a small nit/question on the `hostIP` function, but aside from that, LGTM!

Pre-approving so not to be a blocker later

File: builder/vsphere/common/step_boot_command.go
Comment (tenthirtyam) [reply builder/vsphere/common/step_boot_command.go:230]: Good catch! Resolved in [4f68e62.](https://github.com/hashicorp/packer-plugin-vsphere/commit/4f68e62dea97655f36f46919769bd41c09fffe6d).

**Pull request merged**
